Add Honda Odyssey 2023

tskho opened the pull request:
**Car**
2023 Honda Odyssey

**Route**
d7233a428eb7d0b5/00000017--11cdf651c0

Special thanks to @csouers, his attempt has helped a lot.

Thanks a bunch to @jyoung8607 for invaluable support on this PR.

jyoung8607 commented on selfdrive/car/honda/values.py:
```diff
@@ -163,37 +168,56 @@ class CAR(Platforms):
+    flags=HondaFlags.BOSCH_ALT_BRAKE,
+  )
+  HONDA_ODYSSEY_BOSCH = HondaBoschPlatformConfig(
+    [HondaCarDocs("Honda Odyssey 2021-2023",
```
Change this to the standard docs format of YYYY-YY. This accounts for a couple of the CI failures.

```suggestion
    [HondaCarDocs("Honda Odyssey 2021-23",
```

jyoung8607 commented on selfdrive/car/honda/values.py:
```diff
@@ -163,37 +168,56 @@ class CAR(Platforms):
     [HondaCarDocs("Acura RDX 2019-22", "All", min_steer_speed=3. * CV.MPH_TO_MS)],
-    CarSpecs(mass=4068 * CV.LB_TO_KG, wheelbase=2.75, steerRatio=11.95, centerToFrontRatio=0.41, tireStiffnessFactor=0.677),  # as spec
+    CarSpecs(mass=4068 * CV.LB_TO_KG, wheelbase=2.75, steerRatio=11.95, centerToFrontRatio=0.41,
+             tireStiffnessFactor=0.677),  # as spec
+    dbc_dict('acura_rdx_2020_can_generated', None),
+    flags=HondaFlags.BOSCH_ALT_BRAKE,
+  )
+  HONDA_ODYSSEY_BOSCH = HondaBoschPlatformConfig(
+    [HondaCarDocs("Honda Odyssey 2021-2023",
+                  "All",
+                  min_steer_speed=3. * CV.MPH_TO_MS)],
+    CarSpecs(
+      mass=4535 * CV.LB_TO_KG,
+      wheelbase=3.,
+      steerRatio=14.35,
+      centerToFrontRatio=0.45,
+      tireStiffnessFactor=0.677),  # as spec
     dbc_dict('acura_rdx_2020_can_generated', None),
     flags=HondaFlags.BOSCH_ALT_BRAKE,
   )
   HONDA_INSIGHT = HondaBoschPlatformConfig(
     [HondaCarDocs("Honda Insight 2019-22", "All", min_steer_speed=3. * CV.MPH_TO_MS)],
-    CarSpecs(mass=2987 * CV.LB_TO_KG, wheelbase=2.7, steerRatio=15.0, centerToFrontRatio=0.39, tireStiffnessFactor=0.82),  # as spec
+    CarSpecs(mass=2987 * CV.LB_TO_KG, wheelbase=2.7, steerRatio=15.0, centerToFrontRatio=0.39,
+             tireStiffnessFactor=0.82),  # as spec
```
Looks like your editor tried to be helpful and reformat lines it thought were too long. This change (and a whole bunch of others, this is just an example) need to be reverted. The only changes in the PR should be those necessary to add support for your car.

jyoung8607 reviewed (changes requested):
Once you fix the model-year in CarDocs and regen CARS.md, two of the CI test failures should be fixed. A third test, the car models CI test, is expected to fail until comma uploads your test route into the CI bucket on their side. They'll take care of that when they review.

The whitespace/formatting changes won't stop CI from passing, but they will stop comma from reviewing and merging, so please go ahead and revert those as well.

The `.gitignore` and VScode config JSON changes need to be reverted as well.

tskho committed:
Add Honda Odyssey 2023

Test Route:
d7233a428eb7d0b5/00000017--11cdf651c0

tskho commented:
Prior art here: https://github.com/commaai/openpilot/discussions/2349

jyoung8607 commented on selfdrive/car/honda/carstate.py:
```diff
@@ -206,6 +206,12 @@ def update(self, cp, cp_cam, cp_body):
         ret.cruiseState.standstill = acc_hud["CRUISE_SPEED"] == 252.
 
         conversion = get_cruise_speed_conversion(self.CP.carFingerprint, self.is_metric)
+        # TODO: finish this part
+        # print(f'radar/acc {int(acc_hud["IMPERIAL_UNIT"])} - meter {int(cp.vl["CAR_SPEED"]["IMPERIAL_UNIT"])}')
```
What's going on here? Does this still need work?

jyoung8607 commented on selfdrive/car/honda/fingerprints.py:
```diff
@@ -395,6 +395,46 @@
     ],
   },
+  CAR.HONDA_ODYSSEY_BOSCH: {
+    (Ecu.programmedFuelInjection, 0x18da10f1, None): [
+      b'37805-5MR-C730\x00\x00',  # 2021 Touring
+      b'37805-5MR-C910\x00\x00',  # 2022 Touring
+      b'37805-5MR-C920\x00\x00',  # 2022 Elite
+      b'37805-5MR-C110\x00\x00',  # 2023 Sport
+    ],
```
Remove this block: `programmedFuelInjection` is in [extra_ecus](https://github.com/commaai/openpilot/blob/7de2aac3c93caa20213894e574f06bd45f789335/selfdrive/car/honda/values.py#L308-L316) for Honda, not for fingerprinting.

This and `combinationMeter` are the cause of your build_release CI failure.

```suggestion
```

jyoung8607 commented on selfdrive/car/honda/fingerprints.py:
```diff
@@ -395,6 +395,46 @@
+      b'57114-THR-A230\x00\x00',
+      b'57114-THR-A240\x00\x00',  # 2022 Elite, 2023 Sport
+    ],
+    (Ecu.combinationMeter, 0x18da60f1, None): [
+      b'78109-THR-AS10\x00\x00',  # 2021 Touring
+      b'78109-THR-AT10\x00\x00',  # 2021 ?, 2022 Elite
+      b'78109-THR-CT10\x00\x00',  # 2022 Touring
+      b'78109-THR-AR10\x00\x00',  # 2021 EX-L
+      b'78109-THR-AQ10\x00\x00',  # 2022 EX
+      b'78109-THR-AW10\x00\x00',  # 2023 Sport
+    ],
```
Remove `combinationMeter` for the same reasons as `programmedFuelInjection`.

```suggestion
```

jyoung8607 commented on selfdrive/car/honda/fingerprints.py:
```diff
@@ -395,6 +395,46 @@
   },
+  CAR.HONDA_ODYSSEY_BOSCH: {
+    (Ecu.programmedFuelInjection, 0x18da10f1, None): [
+      b'37805-5MR-C730\x00\x00',  # 2021 Touring
+      b'37805-5MR-C910\x00\x00',  # 2022 Touring
+      b'37805-5MR-C920\x00\x00',  # 2022 Elite
+      b'37805-5MR-C110\x00\x00',  # 2023 Sport
```
Remove the trailing comments on this and all other firmware, they're not really possible to maintain.

```suggestion
      b'37805-5MR-C730\x00\x00',
      b'37805-5MR-C910\x00\x00',
      b'37805-5MR-C920\x00\x00',
      b'37805-5MR-C110\x00\x00',
```

jyoung8607 commented on selfdrive/car/honda/interface.py:
```diff
@@ -161,10 +161,12 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long, docs):
       ret.lateralTuning.pid.kpV, ret.lateralTuning.pid.kiV = [[0.28], [0.08]]
       if candidate == CAR.HONDA_ODYSSEY_CHN:
         ret.lateralParams.torqueBP, ret.lateralParams.torqueV = [[0, 32767], [0, 32767]]  # TODO: determine if there is a dead zone at the top end
+      elif not ret.openpilotLongitudinalControl:
+        ret.lateralParams.torqueBP, ret.lateralParams.torqueV = [[0, 3840], [0, 3840]]  # clipped by radar
```
This may be a little too clever. I had to take time to read other Honda car port code to be sure this couldn't affect HONDA_ODYSSEY, and it might not be robust in the face of new Honda car ports being added. It would be clearer to make this conditional on HONDA_ODYSSEY_BOSCH and keep the comment.

```suggestion
      elif candidate == CAR.HONDA_ODYSSEY_BOSCH:
        ret.lateralParams.torqueBP, ret.lateralParams.torqueV = [[0, 3840], [0, 3840]]  # clipped by radar
```

jyoung8607 reviewed (changes requested):
As a general comment, comma generally wants to see verifiable, traceable evidence of where every fingerprint entry comes from. You've got several in here beyond your own vehicle. At a minimum, comma will want to know routes or at least dongle IDs those firmware entries came from. It would make review faster if you stripped this PR down to your own vehicle's firmware and followed up with a firmware batch PR after this was merged.

tskho commented:
Comma, please upload route  d7233a428eb7d0b5/00000005--4540b82b5b to CI store.

adeebshihadeh commented:
> d7233a428eb7d0b5/00000005--4540b82b5b

Done!

adeebshihadeh commented:
You'll need to upload the rlogs for that route first.

tskho commented:
> You'll need to upload the rlogs for that route first.

If this means Log Upload from connect portal, then I just did that.

tskho commented:
Potential issue discovered: miles on the route number reported on useradmin portal to are too much for the route. I drove for 25 minutes, I doubt I could really do 53 miles in that time in 2023 Honda Odyssey during weekday morning. Actual distance is far less. I wonder how is that calculated and if these car values somehow affect miles counter.

tskho commented on selfdrive/car/honda/carstate.py:
```diff
@@ -206,6 +206,10 @@ def update(self, cp, cp_cam, cp_body):
 
         conversion = get_cruise_speed_conversion(self.CP.carFingerprint, self.is_metric)
+        if self.CP.carFingerprint == CAR.HONDA_ODYSSEY_BOSCH:
+          conversion = CV.MPH_TO_MS if acc_hud['IMPERIAL_UNIT'] else CV.KPH_TO_MS
```
I might need a new test route after this change was applied

tskho replied:
I don't think miles calculator is affected by this change (meaning, still doesn't show right values)

tskho commented:
The only failing check is 

```
Route not on CI bucket. This is fine to fail for WIP car ports, just let us know and we can upload your routes to the CI bucket.
```

Please add test route `d7233a428eb7d0b5/00000017--11cdf651c0` to the CI bucket

tskho commented:
Rebased

sshane commented:
There is a lot of code missing from the 2021-22 port: https://github.com/commaai/openpilot/pull/25450 @csouers do you know if we need it still/any new discoveries made to simplify the port?

sshane commented on selfdrive/car/honda/carstate.py:
```diff
@@ -206,6 +206,10 @@ def update(self, cp, cp_cam, cp_body):
+        if self.CP.carFingerprint == CAR.HONDA_ODYSSEY_BOSCH:
+          conversion = CV.MPH_TO_MS if acc_hud['IMPERIAL_UNIT'] else CV.KPH_TO_MS
+        else:
+          conversion = get_cruise_speed_conversion(self.CP.carFingerprint, self.is_metric)
```
this is the same as the line above. why can't `get_cruise_speed_conversion` handle this?

tskho committed:
Honda Odyssey 2023 LKAS is speed limited

The vehicle is traveling between about 45 and 90 mph (72 and 145 km/h).
https://owners.honda.com/utility/download?path=/static/pdfs/2023/Odyssey/2023_Odyssey_Lane_Keeping_Assist_System.PDF

csouers commented:
> There is a lot of code missing from the 2021-22 port: #25450 @csouers do you know if we need it still/any new discoveries made to simplify the port?

This pr looks to be a good base to build on later.

Implementing better alerting for the steering state would probably be a good idea. A common case would be that the radar is filtering out OP's steering requests due to limitations such as low speed and/or lo/hi wiper mode. The stock LKAS limitations are implemented in the radar's firmware and still apply with OP doing lateral-only. Would have to check the radar's steer request output or the EPS' control state to know. The EPS will not report fault for this.

A low steer speed alert isn't yet implemented for Honda so that would probably have to happen too. I have a PR fixing up the minSteerSpeeds for Honda which would apply to this. I'll rebase that later.

tskho commented:
@csouers I found a flag we can use to detect when steering is filtered. However, I'm interested in what you are saying about wipers. Just today this branch was steering me through rain with wipers on high, no problem at all. OP was more comfortable than me.

csouers commented:
Oh cool. That's good to know. I've seen drives on the earlier models where it would consistently disengage. I'm surprised Honda would have made that change.

sshane commented:
We've moved the car interfacing code to our [`opendbc`](https://github.com/commaai/opendbc) repository, which is now the new home for car ports and fingerprints. Please re-open your pull request against [`opendbc`](https://github.com/commaai/opendbc) at your convenience by using this command below. This will transform all changes under `selfdrive/car/` to `opendbc_repo/opendbc/car/`. Make sure you have initialized submodules beforehand and have checked out the latest `opendbc` commit.

```bash
PR_NUMBER=33045
curl -L https://github.com/commaai/openpilot/pull/$PR_NUMBER.patch | sed -e 's/selfdrive\/car/opendbc_repo\/opendbc\/car/g' | git apply -v --reject
```

Simply replace the PR number with your own. Once done, add the files, fix any conflicts, and open a new PR. Alternatively, you may start a new PR from scratch if that is easier for you.
